File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        
        
        q = deque()
        res = []
        
        for idx, num in enumerate(nums):
            
            while q and q[-1][0] < num:
                q.pop()
            
            q.append((num, idx))
            
            if idx - q[0][1] == k:
                q.popleft()
                
            
            #Dont append res
            if idx + 1 < k:
                continue
            
            res.append(q[0][0])
        
        return res
        

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        deque<pair<int, int>> dq;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            int num = nums[i];
            while (!dq.empty() && dq.back().first < num) dq.pop_back();

            dq.push_back({num, i});

            if (i - dq.front().second >= k) dq.pop_front();

            if (i + 1 >= k) res.push_back(dq.front().first);
        }

        return res;
    }
};